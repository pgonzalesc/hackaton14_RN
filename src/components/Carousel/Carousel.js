import React, {useState, useEffect} from 'react';
import {
  View,
  Text,
  StyleSheet,
  Dimensions,
  FlatList,
  Animated,
} from 'react-native';
import CarouselItem from './CarouselItem';

let flatList;
const infiniteScroll = (datalist) => {
  const numberOfData = datalist.length;
  let scrollValue = 0;
  let scrolled = 0;
  setInterval(() => {
    scrolled++;
    if (scrolled < numberOfData) {
      scrollValue = scrollValue + width;
    } else {
      scrollValue = 0;
      scrolled = 0;
    }
    this.flatList.scrollToOffset({animated: true, offset: scrollValue});
  }, 5000);
};
const {height, width} = Dimensions.get('window');

const styles = StyleSheet.create({
  dot: {
    flex: 0.05,
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
  },
});

const AnimatedFlatList = Animated.createAnimatedComponent(FlatList);

const Carousel = ({data}) => {
  const scrollX = new Animated.Value(0);
  let position = Animated.divide(scrollX, width);
  const [dataList, setDataList] = useState(data);

  useEffect(() => {
    setDataList(data);
    infiniteScroll(dataList);
  }, []);
  if (data && data.length) {
    return (
      <View style={{flex: 1}}>
        <AnimatedFlatList
          ref={(flatList) => {
            this.flatList = flatList;
          }}
          style={{flex: 0.95}}
          data={data}
          keyExtractor={(item, index) => 'key' + index}
          horizontal
          scrollEnabled
          snapToAlignment="center"
          scrollEventThrottle={16}
          decelerationRate="fast"
          showsHorizontalScrollIndicator={false}
          renderItem={(item) => {
            return <CarouselItem item={item.item} />;
          }}
          onScroll={Animated.event(
            [{nativeEvent: {contentOffset: {x: scrollX}}}],
            {
                //listener: (event) => console.log(event),
                useNativeDriver: true,
            },
          )}
        />
        <View style={styles.dot}>
          {data.map((_, i) => {
            let opacity = position.interpolate({
              inputRange: [i - 1, i, i + 1],
              outputRange: [0.3, 1, 0.3],
              extrapolate: 'clamp',
            });
            return (
              <Animated.View
                key={i}
                style={{
                  opacity,
                  height: 10,
                  width: 10,
                  backgroundColor: '#212121',
                  margin: 8,
                  borderRadius: 5,
                }}
              />
            );
          })}
        </View>
      </View>
    );
  }
  return null;
};

export default Carousel;
