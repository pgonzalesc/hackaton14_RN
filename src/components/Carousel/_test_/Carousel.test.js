import React from 'react';
import { } from 'react-native';
import {shallow} from 'enzyme';
import Carousel from '../Carousel';
import {carouselData} from '../../../data/carouselData';

jest.mock('react-native/Libraries/Animated/src/NativeAnimatedHelper');

describe('Pruebas en <Carousel/>', () => {
    let wrapper = shallow(<Carousel data={carouselData}/>);

    beforeEach(()=>{
        wrapper = shallow(<Carousel data={carouselData}/>);
    });

    test('Debe de mostrar <Carousel/> correctamente', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('Debe de renderizar', () => {
        expect(wrapper.exists()).toBe(true);
    });

    test('Debe de coincidir los props', () => {
        const data = wrapper.props().children;
        expect(data[0].props.data).toEqual(carouselData);
    });
});
