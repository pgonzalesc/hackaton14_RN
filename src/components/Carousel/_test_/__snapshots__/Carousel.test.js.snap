// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <Carousel/> Debe de mostrar <Carousel/> correctamente 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <ForwardRef(AnimatedComponentWrapper)
    data={
      Array [
        Object {
          "description": "",
          "id": 1,
          "image": "https://images.unsplash.com/photo-1559141680-d0bd7bc5af84?ixid=MXwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHw%3D&ixlib=rb-1.2.1&auto=format&fit=crop&w=192&q=80",
          "title": "50% Discount for all Purchases this Morning",
        },
        Object {
          "description": "",
          "id": 2,
          "image": "https://images.unsplash.com/photo-1485365435483-337eb45ec196?ixid=MXwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHw%3D&ixlib=rb-1.2.1&auto=format&fit=crop&w=1050&q=80",
          "title": "50% Discount for all Purchases this Morning",
        },
        Object {
          "description": "",
          "id": 3,
          "image": "https://images.unsplash.com/photo-1459257868276-5e65389e2722?ixid=MXwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHw%3D&ixlib=rb-1.2.1&auto=format&fit=crop&w=1050&q=80",
          "title": "50% Discount for all Purchases this Morning",
        },
        Object {
          "description": "",
          "id": 4,
          "image": "https://images.unsplash.com/photo-1448316998199-a2aaf2b5d4a6?ixid=MXwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHw%3D&ixlib=rb-1.2.1&auto=format&fit=crop&w=1050&q=80",
          "title": "50% Discount for all Purchases this Morning",
        },
      ]
    }
    decelerationRate="fast"
    horizontal={true}
    keyExtractor={[Function]}
    onScroll={[Function]}
    renderItem={[Function]}
    scrollEnabled={true}
    scrollEventThrottle={16}
    showsHorizontalScrollIndicator={false}
    snapToAlignment="center"
    style={
      Object {
        "flex": 0.95,
      }
    }
  />
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 0.05,
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <ForwardRef(AnimatedComponentWrapper)
      key="0"
      style={
        Object {
          "backgroundColor": "#212121",
          "borderRadius": 5,
          "height": 10,
          "margin": 8,
          "opacity": 1,
          "width": 10,
        }
      }
    />
    <ForwardRef(AnimatedComponentWrapper)
      key="1"
      style={
        Object {
          "backgroundColor": "#212121",
          "borderRadius": 5,
          "height": 10,
          "margin": 8,
          "opacity": 0.3,
          "width": 10,
        }
      }
    />
    <ForwardRef(AnimatedComponentWrapper)
      key="2"
      style={
        Object {
          "backgroundColor": "#212121",
          "borderRadius": 5,
          "height": 10,
          "margin": 8,
          "opacity": 0.3,
          "width": 10,
        }
      }
    />
    <ForwardRef(AnimatedComponentWrapper)
      key="3"
      style={
        Object {
          "backgroundColor": "#212121",
          "borderRadius": 5,
          "height": 10,
          "margin": 8,
          "opacity": 0.3,
          "width": 10,
        }
      }
    />
  </View>
</View>
`;
